From 8607a15aba0b9f610efa69cc74fa4a178fd2f8ce Mon Sep 17 00:00:00 2001
From: Ewan Mellor <ewan.mellor@citrix.com>
Date: Tue, 1 Nov 2011 11:54:59 -0700
Subject: [PATCH] Bug 884863: nova logs everything to syslog twice

Fix double-registration of the syslog handler.
NovaRootLogger.setup_from_flags is called twice from log.setup() -- once
through NovaRootLogger.__init__ and once through reset().  setup_from_flags
wasn't idempotent, so this resulted in the syslog handler being registered
twice.

Rather than fix the twisty-turny maze that is setup(), I've opted to make
setup_from_flags idempotent in this regard, by always unregistering the
syslog handler before doing anything else.

(cherry picked from commit a807c57f5b61cb25a6f5c140a97ed744ab0a70dd)

Change-Id: I59ad61751e1a19d2cbb73dc1deea9c708d4c5032
---
 nova/log.py |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/nova/log.py b/nova/log.py
index 86f1570..23b83ab 100644
--- a/nova/log.py
+++ b/nova/log.py
@@ -249,11 +249,12 @@ class NovaRootLogger(NovaLogger):
     def setup_from_flags(self):
         """Setup logger from flags."""
         global _filelog
+        if self.syslog:
+            self.removeHandler(self.syslog)
+            self.syslog = None
         if FLAGS.use_syslog:
             self.syslog = SysLogHandler(address='/dev/log')
             self.addHandler(self.syslog)
-        elif self.syslog:
-            self.removeHandler(self.syslog)
         logpath = _get_log_file_path()
         if logpath:
             self.removeHandler(self.streamlog)
-- 
1.7.6.5

